### ISSUE TITLE
some questions in trainer.py

### ISSUE THREAD

**Issue by Terrygth** [pojtanut-mano/thai_stock_prediction_gnn#1]
Hello! 
I think the "train_epoch()" is the real part where you train and update the model. But why you do "self.model.forward()" in the "train_classifier" near 52 lines?

Maybe  "self.model.train()" and "self.model.eval()" make  sense, but can we caculating train_acc and train_f1 in the "train_epoch"?

I think you do "forward" twice in one epoch, is that?

**Comment by pojtanut-mano** [comment]
Hello Terrygth!

`train_epoch` method is the part where the model is trained and updated in each epoch.

As for the question of whether we can calculate the metrics in `train_epoch`, I think yes, we can do that. The concern that I had at the moment of writing this code was quite trivial. I thought that when we pass one epoch for training; In each minibatch, the model gets updated and parameters change, so the accuracy measured at each minibatch would not be a good representative for that epoch. However, I guess we could say that it affects the run-time by a margin and could be done in one step in `train_epoch`.

If I can remember it correctly, yes, in one epoch, the code runs forward twice, once for training the model and once for acquiring the loss number.

Feel free to drop more questions if you are confused with any part. I will try to answer the best that I could.
